Add a Physics for linear FV + mixture model

GiudGiud opened the pull request:
Requires #29177 and maybe #29516 to try out transients (or some other way to get a converged solution)

One difficulty is likely that the two phase equation needs to be an "active" scalar, not a passive one, and be solved inside the pressure-velocity-energy picard iteration.
I added code for that. I ll revisit this PR once the prereqs are merged

GiudGiud committed:
Add a Physics for linear finite volume for two phase flow
refs #22637 #29175

GiudGiud committed:
Adapt slip material and use of slip velocity to linear variables
refs #22637 #29175

GiudGiud committed:
Add parameter setting routines for two phase mixtures to be shared between linear and nonlinear implementations
refs #22637 #29175

GiudGiud committed:
Add an input for lid-driven + linear FV
Currently a steady solve with only velocity-pressure, not converging. Try transient
WIP
refs #22637 #29175

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2617091/), step Clang format on 213f26e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2974c54c5269c9b763547954236d33e3a55a8710`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2617091/), step Fixup commit check on 213f26e wanted to post the following:

The following commit(s) contain fixups:

- e246047cd0: fixup! Add a Physics for linear finite volume for two phase flow

GiudGiud committed:
Add support for active scalars in linear fv segregated assembly solve objects
refs #22637 #29175

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2617448/), step Clang format on 8248e8f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2974c54c5269c9b763547954236d33e3a55a8710`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2617495/), step Docs: sync website on c7c56d2 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29614/site)

This comment will be updated on new commits.

GiudGiud committed:
Avoid thinking the pressure pin works in linear finite volume through the Physics

GiudGiud committed:
Add linear system summary output to header and mesh-changed console output
closes #29790

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2664725/), step Clang format on 4586129 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 32662f974835a45bece8bcfc734f7bf5d5ec5ea1`

GiudGiud committed:
Request cell gradients when using the slip velocities
refs #22637 #29175

GiudGiud committed:
Add face side element to linear fv advection so that advection by functor material properties
( = piecewisebyblock lambda functors) can work
Add an assert for having a flow physics
refs #22637 #29175

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2664853/), step Clang format on f8dc1b6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 32662f974835a45bece8bcfc734f7bf5d5ec5ea1`

GiudGiud committed:
Remove transient term for now from slip velocity

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2664904/), step Clang format on 63f46dd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 32662f974835a45bece8bcfc734f7bf5d5ec5ea1`

GiudGiud committed:
Output exceptions in linear system

tanoret commented:
Once the PR is ready, please consider adding a technical reviewer as per our suggested merge standards to NS. Thanks!

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVTwoPhaseMixturePhysics.C:
```diff
@@ -0,0 +1,313 @@
+                 "' should be 'cp_mixture'");
+  }
+  if (_flow_equations_physics)
+  {
+    if (_flow_equations_physics->densityName() != "rho_mixture")
+      mooseError("Density name should for Physics ,",
+                 _flow_equations_physics->name(),
+                 "' should be 'rho_mixture'");
+  }
```
```suggestion
    if (_flow_equations_physics->densityName() != "rho_mixture")
      mooseError("Density name should for Physics ,",
                 _flow_equations_physics->name(),
                 "' should be 'rho_mixture'");
```

grmnptr reviewed (changes requested):
Nice, just a few comments!

GiudGiud committed:
Address Peter's review

- reduce bracket count
- better messages to the user
- avoid normalizing by a number below 1e-10 (previously 1e-16)

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

GiudGiud commented on modules/navier_stokes/src/executioners/LinearAssemblySegregatedSolve.C:
```diff
@@ -409,6 +501,27 @@ LinearAssemblySegregatedSolve::solve()
+    // and velocity
+    // We wait one iteration for the slip velocity to be updated (phase fraction active scalar)
+    // NOTE: we are lagging slip velocity by 1 iteration
+    if (_has_active_scalar_systems && simple_iteration_counter > 1)
```
is the lag preferred over trying to get the slip velocity updated straight away? this 1 iteration difference will remain throughout

GiudGiud replied:
the lagging was removed

GiudGiud commented on modules/navier_stokes/src/executioners/LinearAssemblySegregatedSolve.C:
```diff
@@ -312,7 +400,7 @@ LinearAssemblySegregatedSolve::solveAdvectedSystem(const unsigned int system_num
 
   // We compute the normalization factors based on the fluxes
-  Real norm_factor = NS::FV::computeNormalizationFactor(solution, mmat, rhs);
+  Real norm_factor = std::max(1e-10, NS::FV::computeNormalizationFactor(solution, mmat, rhs));
```
will do

GiudGiud commented on modules/navier_stokes/include/linearfvkernels/LinearFVScalarAdvection.h:
```diff
@@ -49,6 +49,16 @@ class LinearFVScalarAdvection : public LinearFVFluxKernel
   /// matrix and right hand side contribution
   Real _volumetric_face_flux;
 
+  /// slip velocity in direction x
+  const Moose::Functor<ADReal> * const _u_slip;
+  /// slip velocity in direction y
+  const Moose::Functor<ADReal> * const _v_slip;
+  /// slip velocity in direction z
+  const Moose::Functor<ADReal> * const _w_slip;
+
+  /// Whether to use an additional slip velocity to compute the face flux
+  bool _add_slip_model;
+
```
if I do that I cant do `  WCNSLinearFVScalarTransportPhysics::addFVKernels();` in the lienarFV2phasemixutrephysics

GiudGiud replied:
also that's not what we did for nonlienar FV

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2686576/), step Clang format on b6d1d9d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cd66266a7dec6c2ec22f5902dba3e2ab6eeb8b19`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2687164/), step Generate coverage on 1dd4433 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>cd6626</th>
      <th colspan=4>#29614 1dd443</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.29%</td>
      <td>85.29%</td>
      <td>-0.00%</td>
      <td>73.91%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108457</td>
      <td>108466</td>
      <td>+9</td>
      <td>17</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18702</td>
      <td>18711</td>
      <td>+9</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29614/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29614/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>cd6626</th>
      <th colspan=4>#29614 1dd443</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.83%</td>
      <td>83.97%</td>
      <td>-0.86%</td>
      <td>25.08%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17777</td>
      <td>17827</td>
      <td>+50</td>
      <td>75</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3178</td>
      <td>3403</td>
      <td>+225</td>
      <td>224</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29614/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29614/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29614/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 73.91% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 25.08% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
any preferred test case on this? It s only 1 model for mixture. no solidification or any refinement.
and any volunteer for technical review? (once I clean up the tests etc, not now)

GiudGiud committed:
Address non-unity test failure by adding header
Add missing doc and syntax page
refs #22637 #29175

GiudGiud commented:
@tanoret I can try to reproduce the rayleigh-bernard convection input with Physics & linear FV. Is that enough ? Is there a particular flow test bed input you would like to see?

GiudGiud committed:
Add two phase rayleigh bernard convection test
refs #22637 #29175

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2690546/), step Clang format on bfd72a5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c8ca621a331f92c50b52b8db2eb5c4f89fc38ba1`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2690597/), step Clang format on 77536ed wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c8ca621a331f92c50b52b8db2eb5c4f89fc38ba1`

GiudGiud commented:
The results are not a great match for the two test cases between linear and nonlinear.

GiudGiud committed:
Add support for the phase change interface term in two phase mixtures

GiudGiud committed:
Avoid executing auxkenrels twice

GiudGiud committed:
Add support for the drift flux model in the two phase mixture physics

GiudGiud committed:
Improve parameter groups in flow physics

GiudGiud committed:
Add a test for channel flow + drift flux + linearFV + physics

GiudGiud committed:
Add a kernel for the friction term for the linear FV discretization

GiudGiud committed:
Add support for friction kernels to the linear FV flow physics

GiudGiud committed:
Add a test for the new linear fv friction kernel

GiudGiud committed:
Remove iteration lag in active scalar solves

GiudGiud committed:
Add support for not setting energy wall BCs on every momentum wall boundary
refs #29955

grmnptr commented on modules/navier_stokes/src/physics/WCNSLinearFVTwoPhaseMixturePhysics.C:
```diff
@@ -0,0 +1,343 @@
+void
+WCNSLinearFVTwoPhaseMixturePhysics::addPhaseInterfaceTerm()
+{
+  // Re-create the phase interface term from existing kernels
```
```suggestion
  // Recreate the phase interface term from existing kernels
```

grmnptr commented on modules/navier_stokes/src/utils/SegregatedSolverUtils.C:
```diff
@@ -173,7 +173,7 @@ computeNormalizationFactor(const NumericVector<Number> & solution,
   // avoid normalizing with 0.
   // TODO: Would be nice to see if we can do l1 norms in the linear solve.
-  return (A_times_average_solution->l2_norm() + libMesh::TOLERANCE * libMesh::TOLERANCE);
+  return (A_times_average_solution->l2_norm() + 1e-2 * libMesh::TOLERANCE);
```
```suggestion
  return (A_times_average_solution->l2_norm() + 1e-8);
```

grmnptr commented on modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/segregated/mixture_physics_diff.cmp:
```diff
@@ -0,0 +1,12 @@
+# No GLOBAL VARIABLES
+
+ELEMENT VARIABLES
+	vel_x       relative 2e-1
+	vel_y       relative 5e-1
+	pressure    relative 6e-2
+	phase_2     relative 1e-2
+	phase_1_out relative 1e-2
```
Could you just gold them separately? Obviously the two are not the same.

GiudGiud replied:
1% on the phase over the whole domain is pretty good?

grmnptr commented on modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/segregated/rayleigh-bernard-two-phase-physics.i:
```diff
@@ -0,0 +1,223 @@
+  []
+[]
+
+[AuxVariables]
+  [U]
+    order = CONSTANT
+    family = MONOMIAL
+    fv = true
+  []
+[]
+
+[AuxKernels]
+  [mag]
+    type = VectorMagnitudeAux
+    variable = U
+    x = vel_x
+    y = vel_y
+  []
+[]
```
Do we need this?

GiudGiud replied:
deleted

grmnptr reviewed (changes requested):
Nice! Just a few questions.

GiudGiud committed:
Address Peter's review

- specialize friction kernel doco for Darcy (only model)
- fix sign in drift flux model
- re-add add_scalar_eq parameter to be able to use it in doco

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

GiudGiud commented on modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTwoPhaseMixturePhysics.md:
```diff
@@ -0,0 +1,60 @@
+# Navier Stokes Two Phase Mixture using a Linear Finite Volume discretization / WCNSLinearFVTwoPhaseMixturePhysics
```
the actual object names arent present in the input so I chose to be descriptive in the titles instead

GiudGiud commented on modules/navier_stokes/include/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.h:
```diff
@@ -0,0 +1,91 @@
+ * Adds drift flux kernel coming for two-phase mixture model for the linear finite volume
+ * discretization
+ */
+class LinearWCNSFV2PMomentumDriftFlux : public LinearFVFluxKernel
```
the nonlinear kernels and functor materials use 2P

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearFVMomentumFriction.C:
```diff
@@ -0,0 +1,63 @@
+  InputParameters params = LinearFVElementalKernel::validParams();
+  params.addClassDescription(
+      "Computes a Darcy friction force term on fluid in the "
+      "Navier Stokes i-th momentum equation in Rhie-Chow (incompressible) contexts.");
```
```suggestion
      "Navier Stokes i-th momentum equation.");
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2716780/), step Clang format on 6113b53 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5adf17aea9f3318f9617b10e7be1c26d1a99eca9`

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.C:
```diff
@@ -0,0 +1,192 @@
+
+  Real face_rho_fd;
+  if (on_boundary)
+    face_rho_fd = _rho_d(makeCDFace(*_current_face_info), state) *
+                  _f_d(makeCDFace(*_current_face_info), state);
```
```suggestion
    face_rho_fd = _rho_d(face_arg, state) *
                  _f_d(face_arg, state);
```

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.C:
```diff
@@ -0,0 +1,192 @@
+  if (std::abs(u_old) > 1e-6)
+    return _velocity_interp_coeffs.second * _face_flux * (u / u_old - 1) * _current_face_area;
+  else
+    return _face_flux * _current_face_area;
```
```suggestion
    return - _face_flux * _current_face_area;
```

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.C:
```diff
@@ -0,0 +1,192 @@
+  if (std::abs(u_old) > 1e-6)
+    return _velocity_interp_coeffs.second * _face_flux * (u / u_old - 1) * _current_face_area;
+  else
+    return _face_flux * _current_face_area;
```
good point the two if clauses are not consistent here

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.C:
```diff
@@ -0,0 +1,192 @@
+  if (std::abs(u_old) > 1e-6)
+    return _velocity_interp_coeffs.first * _face_flux * (u / u_old - 1) * _current_face_area;
+  else
+    return _face_flux * _current_face_area;
```
```suggestion
    return -_face_flux * _current_face_area;
```

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.C:
```diff
@@ -0,0 +1,192 @@
+{
+  // Lagging the whole term for now
+  // TODO: make sure this only gets called once, and not once per BC
+  return _face_flux * _current_face_area;
```
```suggestion
  return -_face_flux * _current_face_area;
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/segregated/rayleigh-bernard-two-phase-physics.i:
```diff
@@ -0,0 +1,223 @@
+    fv = true
+  []
+[]
+
+[AuxKernels]
+  [mag]
+    type = VectorMagnitudeAux
+    variable = U
+    x = vel_x
+    y = vel_y
+  []
+[]
```
```suggestion
```

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.C:
```diff
@@ -0,0 +1,192 @@
+
+  Real face_rho_fd;
+  if (on_boundary)
+    face_rho_fd = _rho_d(makeCDFace(*_current_face_info), state) *
+                  _f_d(makeCDFace(*_current_face_info), state);
```
changing now

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2716839/), step Clang format on 5691759 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5adf17aea9f3318f9617b10e7be1c26d1a99eca9`

GiudGiud commented on modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTwoPhaseMixturePhysics.md:
```diff
@@ -0,0 +1,60 @@
+## Equation(s)
+
+This [Physics](Physics/index.md) adds terms to the flow and energy equations to account for the presence of a
+two-phase mixture. If specified with the [!param](/Physics/NavierStokes/TwoPhaseMixtureSegregated/WCNSLinearFVTwoPhaseMixturePhysics/add_phase_transport_equation)
```
ah yeah I tried suppressing the parameter that gets renamed into this param. 
I ll unsuppress it

GiudGiud committed:
Fix documentation failures: bad links and some newton text in linearFV doco
Make expensive test only in debug

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearFVMomentumFriction.C:
```diff
@@ -0,0 +1,106 @@
+  InputParameters params = LinearFVElementalKernel::validParams();
+  params.addClassDescription(
+      "Computes a friction force term on fluid in porous media in the "
+      "Navier Stokes i-th momentum equation in Rhie-Chow (incompressible) contexts.");
```
removed

GiudGiud committed:
Add some CSV output to make a regression test for the channel with the physics

GiudGiud commented:
@tanoret or @MengnanLi91 can you please do the technical review / validation?

MengnanLi91 commented on modules/navier_stokes/doc/content/source/physics/WCNSLinearFVTwoPhaseMixturePhysics.md:
```diff
@@ -0,0 +1,60 @@
+
+The phase advection-diffusion equation is:
+
+!equation
```
This equation doesn't seem right. We don't have a advection-diffusion equation for phase variable in mixture model. Do you get it from other documentation?

GiudGiud replied:
I ll double check the equation but I think we do solve for the other phase advection in the two phase mixture model

MengnanLi91 replied:
We have a advection equation for phase volume/void fraction, but no diffusion. I assume here you mean `INSFVScalarFieldAdvection`, right?

MengnanLi91 replied:
This is different from the scalar field used in mixture model. I think it may use for other models. But this equation definitely doesn't show up in the governing equations of mixture model for sure.

MengnanLi91 replied:
I only saw this parameter from `lid-driven-two-phase-physics.i`, not `lid-driven-two-phase.i`.  I probably miss something here?

GiudGiud replied:
`lid-driven-two-phase.i` does not use the Physics syntax

but it does have diffusion
```
  [phase_2_time]
    type = FVFunctorTimeKernel
    variable = phase_2
  []
  [phase_2_advection]
    type = INSFVScalarFieldAdvection
    variable = phase_2
    u_slip = 'vel_slip_x'
    v_slip = 'vel_slip_y'
  []
  [phase_2_diffusion]
    type = FVDiffusion
    variable = phase_2
    coeff = 1e-3
  []
```

GiudGiud replied:
> This is different from the scalar field used in mixture model.

this could very well be true. I am writing the Physics to replicate what was done with the kernel, I am not starting from the equations and the literature

MengnanLi91 replied:
> `lid-driven-two-phase.i` does not use the Physics syntax
> 
> but it does have diffusion
> 
> ```
>   [phase_2_time]
>     type = FVFunctorTimeKernel
>     variable = phase_2
>   []
>   [phase_2_advection]
>     type = INSFVScalarFieldAdvection
>     variable = phase_2
>     u_slip = 'vel_slip_x'
>     v_slip = 'vel_slip_y'
>   []
>   [phase_2_diffusion]
>     type = FVDiffusion
>     variable = phase_2
>     coeff = 1e-3
>   []
> ```

I didn't notice this before. I'll do some search about it. Maybe you are right

MengnanLi91 replied:
Now I remember, `lid-driven-two-phase.i` doesn't follow mixture model formulation. It only uses `slip_velocity`, which is a more simple assumption than mixture model. It doesn't have the `drift_flux` kernel in the momentum equation. Both are valid but we should make the documentation more clear on this. I'll do a PR for that

grmnptr commented:
Can we merhe this @tanoret, @MengnanLi91 ?

MengnanLi91 commented:
> Can we merge this @tanoret, @MengnanLi91 ?

Maybe wait a bit? @tanoret mentioned he has something to change for the mixture model.

GiudGiud committed:
Add support for convective BCs in lienar FV heat transfer physics
Fix up functor material logic

GiudGiud committed:
Avoid using weighting interpolation for pressure

GiudGiud committed:
Input file fixups

GiudGiud committed:
Add an extrapolated BC at the inlet

GiudGiud committed:
Regold due to fixing pressure interpolation method in pdrop PP

GiudGiud commented:
Agreement with nonlinear FV is quite good now with these changes
- extrapolated pressure at the inlet in linear FV
- pressure drop computation fix for the pressure face interpolation

GiudGiud commented:
Rebased on #30288 which should be merged first

GiudGiud committed:
Explicitly disallow mesh skewness with buoyancy forces and 2-phase mixture

GiudGiud commented:
@tanoret the linear FV. two phase mixture will now error with skewed meshes. This should address your concern. 

BTW, I ran a few single phase cases on skewed meshes without skewness correction and got really bad results, which turned into really good results with skewness correction. Maybe the check I have here should just go in all our FV code as well to request skewness correction.

GiudGiud commented:
@grmnptr did you want to re-review this?

GiudGiud committed:
Remove flip of BCs in rayleigh bernard case
Fix bug in gravity kernel

GiudGiud commented:
@grmnptr @tanoret last commit is a bug fix on gravity for the flow physics. 
We should make a decision on this PR so this bug fix makes it in regardless.

GiudGiud committed:
Add an option to disable gravity term in slip velocity to match nonlinear solver results
Rego

GiudGiud committed:
Template mixture functor material to allow use with non-AD lienar FV

GiudGiud committed:
Add a helpful log when debugging AD problem setup

GiudGiud committed:
Improve an error message
Add support for using the conversion T & h functors directly in the input file

GiudGiud committed:
Add a test for heated + two phase mixture with linear FV

GiudGiud committed:
Add test without using the new Physics syntax

GiudGiud committed:
Avoid checking time derivatives on an auxiliary temperature
Add issue for ordering problem in test input affected

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2972838/), step Clang format on 4d5c223 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 35394b6db8be5ebcb26fdaf6275671a8032b47a9`

grmnptr reviewed (changes requested):
Nice!

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.C:
```diff
@@ -0,0 +1,191 @@
+    mooseError(
+        "In three dimensions, the w_slip velocity must be supplied using the 'w_slip' parameter");
+
+  // Phase fraction could be a nonlinear variable
+  const auto & fraction_name = getParam<MooseFunctorName>("fraction_dispersed");
+  if (isParamValid("fraction_dispersed") && _fe_problem.hasVariable(fraction_name))
+    addMooseVariableDependency(&_fe_problem.getVariable(_tid, fraction_name));
```
Removing it. I ll add a note

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.C:
```diff
@@ -0,0 +1,191 @@
+void
+LinearWCNSFV2PMomentumDriftFlux::computeFlux()
+{
+  const auto normal = _current_face_info->normal();
```
le problem with auto. changing

GiudGiud commented on modules/navier_stokes/src/utils/SegregatedSolverUtils.C:
```diff
@@ -173,7 +173,7 @@ computeNormalizationFactor(const NumericVector<Number> & solution,
   // avoid normalizing with 0.
   // TODO: Would be nice to see if we can do l1 norms in the linear solve.
-  return (A_times_average_solution->l2_norm() + libMesh::TOLERANCE * libMesh::TOLERANCE);
+  return (A_times_average_solution->l2_norm() + 1e-8);
```
too old right now I dont remember. I can try the old threshold see what went wrong that I had to change it

GiudGiud replied:
reverted and it passed fine

GiudGiud committed:
Address Peter's review

- fix error messages
- avoid copy construct
- merge blocks in no-physics input
- delete comment calling for RC: RC might not be the move for slip velocity

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

Update modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/segregated/channel-drift-flux.i

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

Update modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/segregated/channel-drift-flux.i

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/segregated/channel-drift-flux.i:
```diff
@@ -0,0 +1,406 @@
+mu = 1.0
```
in a way. this is what happens when you output the Physics syntax to a regular input

grmnptr replied:
Could you get them into one block each please?

GiudGiud commented on modules/navier_stokes/src/linearfvkernels/LinearWCNSFV2PMomentumDriftFlux.C:
```diff
@@ -0,0 +1,191 @@
+
+  // Caching the interpolation coefficients so they will be reused for the matrix and right hand
+  // side terms
+  // Note: We should use Rhie Chow coefficients here, because the way we split the flux is:
```
my reasoning was this is the `_velocity_interp_coeffs` so it s used to get the velocity at the face, which is then multiplied by the face flux (face flux is interpolated with the advected interpolation)

in `LinearWCNSFV2PMomentumDriftFlux::computeElemMatrixContribution()` we see this coefficient multiple the face flux (already interpolated) because the term is v_elem/v_old * coeff * face_flux.

imo we should be fine here? we can decide that we do this v_elem / v_old trick with either interpolation schemes

grmnptr replied:
Traditionally, the advecting flux is interpolated with average/rc and the advected quantity uses coeff/(1-coeff). That flux is lagged, while the advected quantity interpolation is treated implicitly. It feels like we have it backwards here. Especially that in the other advection kernels we have it flipped.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2993489/), step Clang format on 5f47a7a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29614/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0215c4d571eff0b3766b500498b9050c07da8afc`

GiudGiud commented:
@grmnptr

grmnptr reviewed (changes requested):
Last two tiny comments.

grmnptr reviewed (approved):
Nice!
